# 第五次作業

In [1]:
# 15.13, 15.19, 15.38, 15.43, 15.80, 15.85

In [2]:
import pandas as pd
import numpy as np
import scipy.stats as stats
import statsmodels.api as sm
import statsmodels.stats.api as sms
import statsmodels.formula.api as smf
import statsmodels.stats.multicomp as smm
import statsmodels
import scipy
from matplotlib import pyplot as plt
%matplotlib inline
# 設定圖形大小; DPI越大圖越大
plt.rcParams["figure.dpi"] = 120

<div style="background-color:#D9D9D9;"><font size="5"> 15.13</font></div>

H0 : pi = ai forall types <br>
H1 : At least 1 pi & ai differs

In [3]:
df = pd.read_excel('Xr15-13.xlsx')
alpha = 0.05
nobs = df.shape[0]
df_v = df.value_counts()
df_v = pd.DataFrame(df_v, columns=['A'])
df_v = df_v.sort_values('Car type')
df_use= df_v.to_numpy()
df_use = df_use.reshape(1,-1)
prob_e = np.array([0.15, 0.25, 0.40, 0.20])
freq_e = nobs * prob_e
freq_o = df_use
freq_o = [i for k in freq_o for i in k]
freq_o

[36, 58, 74, 29]

In [4]:
if np.sum(freq_e < 5) > 0:
    print("Rule of five is not met. ")
else:
    stat, pvalue = stats.chisquare(freq_o, freq_e)
    print("Chi-squared test: statistics = %0.4f, p-value = %0.4f" % (stat, pvalue))

    df = len(freq_o)-1
    crit_value = stats.chi2.ppf(1 - alpha, df)
    print("Critical value = %0.4f (defree of freedom = %d)" % (crit_value, df))

Chi-squared test: statistics = 6.0000, p-value = 0.1116
Critical value = 7.8147 (defree of freedom = 3)


Since 0.1116 > 0.05, there is no sufficient evidence at 5% significance level to reject the null hypothesis. Thus, we can't infer that certain sizes of cars's rate of car accident is different from expected percentage of accidents.

<div style="background-color:#D9D9D9;"><font size="5"> 15.19</font></div>

H0 : pi = ai forall types <br>
H1 : At least 1 pi & ai differs

In [5]:
df = pd.read_excel('Xr15-19.xlsx')
alpha = 0.05
#nobs = df.shape[0]
df_v = np.array([26, 2, 13, 185, 414, 184, 122, 44, 24, 57, 95, 93])
nobs = df_v.sum()
prob_e = np.array([24, 30, 67, 169, 212, 174, 80, 57, 59, 49, 49, 28])*0.001
freq_e = nobs * prob_e
freq_o = df_v
print(freq_e)

[ 30.216  37.77   84.353 212.771 266.908 219.066 100.72   71.763  74.281
  61.691  61.691  35.252]


In [6]:
if np.sum(freq_e < 5) > 0:
    print("Rule of five is not met. ")
else:
    stat, pvalue = stats.chisquare(freq_o, freq_e)
    print("Chi-squared test: statistics = %0.4f, p-value = %0.4f" % (stat, pvalue))
    df = 12-1
    crit_value = stats.chi2.ppf(1 - alpha, df)
    print("Critical value = %0.4f (defree of freedom = %d)" % (crit_value, df))

Chi-squared test: statistics = 347.3334, p-value = 0.0000
Critical value = 19.6751 (defree of freedom = 11)


Since p < 0.05, there is sufficient evidence at 5% significance level to reject the null hypothesis. Thus, we can infer that the tornadoes percentage have changed.

<div style="background-color:#D9D9D9;"><font size="5"> 15.38</font></div>

H0: The first criterion (quit smoke or not) is independent of the second criterion (education).<br>
H1: two variables are dependent.

In [7]:
df = pd.read_excel('Xr15-38.xlsx')
cont1 = pd.crosstab(df["Education"], df["Smoker"])
print("Contingency Table:")
display(cont1)

Contingency Table:


Smoker       1    2
Education          
1           34   23
2          251  212
3          159  248
4           16   57

In [8]:
chi2, p, dof, ex = stats.chi2_contingency(cont1, correction=False)
print("Chi-squared Statistics = %0.4f" % chi2)
print("Degree of freedom = %0.4f" % dof)
print("p-value = %0.4f" % p)
print("Expected Frequency:")
print(ex)
print("all the Expected Frequency is above 5")

Chi-squared Statistics = 41.7645
Degree of freedom = 3.0000
p-value = 0.0000
Expected Frequency:
[[ 26.22  30.78]
 [212.98 250.02]
 [187.22 219.78]
 [ 33.58  39.42]]
all the Expected Frequency is above 5


Since the p-value is smaller than 0.05, we can reject the null hypothesis, the two factors are dependent.<br>
there is sufficient evidence to show that the amount of education is a factor in determining whether a smoker will quit.

<div style="background-color:#D9D9D9;"><font size="5"> 15.43</font></div>

H0: The first criterion (weapon used) is independent of the second criterion (year).<br>
H1: two variables are dependent.

In [9]:
df = pd.read_excel('Xr15-43.xlsx')
df1 = pd.DataFrame(df['Year 2000'].value_counts())
df2 = pd.DataFrame(df['Year 2005'].value_counts())
df3 = pd.DataFrame(df['Year 2010'].value_counts())
res = pd.merge(df1,df2, left_index=True, right_index=True, how='outer')
res = pd.merge(res,df3, left_index=True, right_index=True, how='outer')
res

Year 2000  Year 2005  Year 2010
1.0        161        175        131
4.0        159        166        126
3.0         53         39         27
2.0         36         37         24

In [10]:
chi2, p, dof, ex = stats.chi2_contingency(res, correction=False)
print("Chi-squared Statistics = %0.4f" % chi2)
print("Degree of freedom = %0.4f" % dof)
print("p-value = %0.4f" % p)
print("Expected Frequency:")
print(ex)
print("all the Expected Frequency is above 5")

Chi-squared Statistics = 4.7588
Degree of freedom = 6.0000
p-value = 0.5751
Expected Frequency:
[[168.4329806  171.72751323 126.83950617]
 [162.6622575  165.84391534 122.49382716]
 [ 42.91975309  43.75925926  32.32098765]
 [ 34.98500882  35.66931217  26.34567901]]
all the Expected Frequency is above 5


Since the p-value is larger than 0.05, we can't reject the null hypothesis, there isn't enough evidence to show that the difference of the year is a factor in determining use of weapons in robberies.

<div style="background-color:#D9D9D9;"><font size="5"> 15.80</font></div>

H0 : the data are normally distributed<br>
H1 : it is not

In [11]:
alpha = 0.1
p1 = stats.norm.cdf(-1)
p2 = stats.norm.cdf(0)-p1
p3 = stats.norm.cdf(1) - stats.norm.cdf(0)
p4 = 1-stats.norm.cdf(1)
e = np.array([p1, p2, p3, p4])*50
f = np.array([6, 27, 14, 3])
print(e,"all above 5")
sum_all = 0
for i in range(e.shape[0]):
    #print(e[i],f[i])
    sum_all += ((e[i]-f[i])**2)/e[i]
sum_all
crit_value = stats.chi2.ppf(1 - alpha, 4-1-1)
print("Since",sum_all,">",crit_value, "\n we reject H0 in favor of H1 under 0.1 significance level,\nthe data are not normal.")

[ 7.9327627 17.0672373 17.0672373  7.9327627] all above 5
Since 9.870083556749014 > 4.605170185988092 
 we reject H0 in favor of H1 under 0.1 significance level,
the data are not normal.


<div style="background-color:#D9D9D9;"><font size="5"> 15.85</font></div>

In [12]:
df = pd.read_excel('Xr13-115.xlsx')
df1 = df['Method 1']
d_mean = df1.mean()
d_std = df1.std()
print("Method 1 :",d_mean, d_std)

Method 1 : 8.89 0.43063260959249733


Method 1<br>
H0 : the data follows a normal distribution with mean 8.89 and standard deviation 0.43<br>
H1 : it doesn't <br>
We divide to 2 groups

In [13]:
df1 = df1.sort_values()
df1
#sort to 2 groups:8.89 as thr boundary
fi = [5,5]
ei = [5,5]
chi2 = 0
print("but when computing critical value, notice that the dataframe is 2-2-1<0, can't apply the test")
print("so is the method 2, the experiment design is inappropriate.")

but when computing critical value, notice that the dataframe is 2-2-1<0, can't apply the test
so is the method 2, the experiment design is inappropriate.
